### PR TITLE
vehicles: make shock absorbers actually absorb shock damage

### PR DESCRIPTION
**Pull request by mlangsdorf**
#### Summary
```SUMMARY: Bugfixes "vehicles: make shock absorbers actually absorb shock damage"```

#### Purpose of change
Shock absorbers are supposed to absorb transmitted shock damage from collision, but they're just low grade armor.  Fix them so they aren't armor at all (though they take up the armor location) and absorb transmitted shock damage properly.

#### Describe the solution
Add the SHOCK_ABSORBER flag to shock absorbers, and change vehicle::damage_all() so that parts on the same tile as a
shock absorber take substantially less damage from transmitted shock damage in collisions.

#### Testing
I spawned a solar car and welded a shock absorber to the (-4, 0) frame, then I saved the game.  Running into a concrete barrier in reverse at -20 mph completely wrecked every upgraded solar panel in the rear of the car.

I added the patch and reloaded the game.  This time, running into the concrete barrier in reverse at -20 mph completely wrecked every solar panel in the rear of the car except the one at (-4, 0), which was unscathed.

Repeating the test with a collision speed of -45 mph wrecked all the solar panels again.

### CODE REVIEW

**Comment by kevingranade** [comment]
Wait, so shock absorbers act as shock isolation just for the location where they are installed?
I was imagining them as more like a bumper that protected the whole vehicle if it was at the impact site.

I see that the "info" field for the property is clear about it, might be a good idea to adjust the vehicle part to match.

**Comment by mlangsdorf** [comment]
shock absorbers as bumpers has three problems:
* you only get one armor part, so if you put a shock absorber on the edge, you're not armoring the edge
* shock absorbers become more powerful as your vehicle gets larger, because they protect more parts by attenuating the shock for all parts beyond the part
* I really don't want to try to figure out whether there is a shock absorber between part #4 and part #21; checking whether there is a shock absorber at a tile is easy.

I'll update the JSON_FLAGS.md, of course.

**Comment by nexusmrsep** [comment]
What about crumple zones? Are shock absorber equal to crumple zones or are they something else?

How I imagine crumple zone parts would work: Lets say that `damage_all` is made to parse vehicle parts in a particular order where each tier is a set of parts with 'x' distance from `impact` `point`. This would simulate damage traveling in a wave along a stiff structure of the vehicle (now it's only used to determine damage, but the order of it's application is irrelevant).  If any given part in tier 1 (first pass, parts nearest to the impact zone) has a crumple zone, it can absorb 'x' amount of damage, and that damage could be deducted from the damage that is passed to calculation for tiers 1+, but not in tier 1 since every tile in this set is experiencing same energy from impact simultaneously (even though they will be processed in order) etc.

In other words if the front row of the car tiles has crumple zones built in front each such zone would contribute to protecting the second row of car tiles. This contribution can also be modeled - damage for the crumple zone tile could be actually increased as it is specifically designed for a destructive absorption of impact energy. Result: front of the car smashed to smithereens, but the passenger is intact.

**Comment by mlangsdorf** [comment]
Crumple zones are way beyond the scope of this PR.

**Comment by kevingranade** [comment]
This pull request has been mentioned on **Cataclysm: Dark Days Ahead**. There might be relevant details there:

https://discourse.cataclysmdda.org/t/collision-damage-through-the-roof-intended/25383/2

**Comment by kevingranade** [comment]
This pull request has been mentioned on **Cataclysm: Dark Days Ahead**. There might be relevant details there:

https://discourse.cataclysmdda.org/t/does-shock-absorber-reduce-the-transmission-of-shock-damage/27086/1
